feat: expressjs framework (#784)
canonical/rockcraft: docs/reference/extensions/expressjs-framework.rst

.. _expressjs-framework-reference:

expressjs-framework
-------------------

The ExpressJS extension streamlines the process of building ExpressJS
application rocks.

It facilitates the installation of ExpressJS application dependencies, including
Node and NPM, inside the rock. Additionally, it transfers your project files to
``/app`` within the rock.

The ExpressJS extension is compatible with the ``bare`` and ``ubuntu@24.04``
bases.

Project requirements
====================

There are 3 requirements to be able to use the ``expressjs-framework``
extension:

1. The application should reside in the ``app`` directory.
2. The application should have a ``package.json`` file.
3. The ``package.json`` file should define the ``start`` script.

``parts`` > ``expressjs-framework/install-app`` > ``npm-include-node``
======================================================================

You can use this field to specify the version of Node to be installed. For
example:

.. code-block:: yaml

  parts:
    expressjs-framework/install-app:
      npm-include-node: true
      npm-node-version: 20.12.2

> For more examples of the ``npm-node-version`` key, see the
> `NPM plugin <httpshttps://documentation.ubuntu.com/rockcraft/en/latest/common/
> craft-parts/reference/plugins/npm_plugin/#examples>`_ documentation.

Reviewer: ```suggestion
For more examples of the ``npm-node-version`` key, see :ref:`craft_parts_npm_plugin`.
```

.. _expressjs-framework-reference:

expressjs-framework
-------------------

The ExpressJS extension streamlines the process of building ExpressJS
application rocks.

It facilitates the installation of ExpressJS application dependencies, including
Node and NPM, inside the rock. Additionally, it transfers your project files to
``/app`` within the rock.

.. note::
    The ExpressJS extension is compatible with the ``bare`` and ``ubuntu@24.04``
    bases.

Project requirements
====================

There are 3 requirements to be able to use the ``expressjs-framework``
extension:

1. The application should reside in the ``app`` directory.
2. The application should have a ``package.json`` file.

> 3. The ``package.json`` file should define the ``start`` script.

Author: The `package.json` file with the `start` script comes with the ExpressJS' official `express-generator` tool and is a common script in the NodeJS community to start the server.

> 3. The ``package.json`` file should defined the ``start`` script.

Reviewer 2: ```suggestion
3. The ``package.json`` file should define the ``start`` script.
```

``parts`` > ``expressjs-framework/install-app`` > ``npm-include-node``
======================================================================

You can use this field to specify the version of Node to be installed. For
example:

.. code-block:: yaml

  parts:
    expressjs-framework/install-app:
      npm-include-node: true
      npm-node-version: 20.12.2

For more examples of npm-node-version options, see: https://documentation.\
ubuntu.com/rockcraft/en/1.5.3/common/craft-parts/reference/plugins/npm_plugin/\

> #examples

Reviewer 2: The link doesn't seem to render correctly in the RTD build.
I suggest the following alternative:
```
`Rockcraft | NPM Plugin | Examples <https://documentation.ubuntu.com/rockcraft/en/1.5.3/common/craft-parts/reference/plugins/npm_plugin/#examples>`_
```

> For more examples of npm-node-version options, see: `Rockcraft | NPM Plugin |
> Examples <https://documentation.ubuntu.com/rockcraft/en/1.5.3/common/\
> craft-parts/reference/plugins/npm_plugin/#examples>`_.

Reviewer 2: This link seems to be broken in the RTD rendering. I think the issue is the inclusion of the `\` -- I think the problem will go away if you remove it from the link.
```suggestion
For more examples of npm-node-version options, see: `Rockcraft | NPM Plugin |
Examples <https://documentation.ubuntu.com/rockcraft/en/1.5.3/common/
craft-parts/reference/plugins/npm_plugin/#examples>`_.
```

If you don't customise the version of node, it will be installed from the Ubuntu
package repository.

``parts`` > ``expressjs-framework/runtime:`` > ``stage-packages``
=================================================================

Installing additional runtime packages is currently unsupported due to some
limitations of the Rockcraft tool.

Useful links

> ============

Author: We would be diverging from the patterns used in other frameworks with this change - would this be intended?

> ============

Reviewer: That's fine – those tutorials are undergoing continuous updates.

.. _expressjs-framework-reference:

expressjs-framework
-------------------

The ExpressJS extension streamlines the process of building ExpressJS
application rocks.

It facilitates the installation of ExpressJS application dependencies, including
Node and NPM, inside the rock. Additionally, it transfers your project files to
``/app`` within the rock.

.. note::
    The ExpressJS extension is compatible with the ``bare`` and ``ubuntu@24.04``
    bases.

Project requirements
====================

There are 3 requirements to be able to use the ``expressjs-framework``
extension:

1. The application should reside in the ``app`` directory.
2. The application should have a ``package.json`` file.
3. The ``package.json`` file should defined the ``start`` script.

``parts`` > ``expressjs-framework/install-app`` > ``npm-include-node``
======================================================================

You can use this field to specify the version of Node to be installed. For
example:

.. code-block:: yaml

  parts:
    expressjs-framework/install-app:
      npm-include-node: true
      npm-node-version: 20.12.2

> For more examples of npm-node-version options, see:
> https://documentation.ubuntu.com/rockcraft/en/1.5.3/common/craft-parts/reference/plugins/npm_plugin/#examples

Reviewer 2: ```suggestion
For more examples of npm-node-version options, see the [Rockcraft documentation on the NPM Plugin](https://documentation.ubuntu.com/rockcraft/en/1.5.3/common/craft-parts/reference/plugins/npm_plugin/#examples).
```

If you don't customise the version of node, it will be installed from the Ubuntu
package repository.

``parts`` > ``expressjs-framework/runtime:`` > ``stage-packages``
=================================================================

You can use this key to specify any dependencies required for your ExpressJS
application. In the following example we use it to specify ``libpq-dev``:

.. code-block:: yaml

  parts:
    expressjs-framework/runtime:
      stage-packages:
        # when using the default provided node version, the npm package is
        # required.
        # - npm
        # list required packages or slices for your ExpressJS application below.
        - libpq-dev

When using the NPM plugin installed Node and NPM, additional runtime packages

> is currently unsupported due to an issue with ``lib`` path permission error. See
> https://github.com/canonical/rockcraft/issues/790 for more information.

Reviewer 2: ```suggestion
is currently unsupported due to an issue with ``lib`` path permission error. See
[Issue #790](https://github.com/canonical/rockcraft/issues/790) for more information.
```
